Introduce script detecting affected tests

sculpt0r opened the pull request:
Add script converting file list from git difference between two branches into bender filters.
Add mocha tests.
It could be run via 
```
node src/diff/differ.js targetBranch currentBranch
```


Closes #1

f1ames commented on src/diff/differ.js:
```diff
@@ -0,0 +1,24 @@
+	const benderFilter = benderFilters.join(',');
+
+	console.log(benderFilter);
+});
```
Do you think it would always get all data at once? Or it should be buffered/merged here and processed on `close` event when we are sure nothing more will show up?

sculpt0r replied:
Well technically it is possible that data comes in multiple chunks, so indeed, it will be better to collect all chunks and process them later. :+1:

f1ames commented on src/diff/differ.js:
```diff
@@ -0,0 +1,24 @@
+});
+
+ls.on('error', (error) => {
+	console.log(`error: ${error.message}`, error);
```
Well, I wonder how errors should be handled. Because `console.log` will use std output and the assumption is this script will print there list of filters. So if error is thrown it may break other scripts which relay on this one...

Maybe it should output either `ERROR: message` or list of filters - so it will be easy to check and parent script can act accordingly?

sculpt0r replied:
Well :cry: standard output was only for early verification if this script working somehow. I will wrap this into casual functions, so the Client for this 'module' would only call normal function.

<strike>Therefore casual log to console from the git process should be fine? But I will add proper headers to logs, so there will be visible if error was printed.</strike>

Proper data is now passed to `resolve` or `reject` method.

f1ames commented on src/diff/diffAnalyzer.js:
```diff
@@ -0,0 +1,81 @@
+	return 'path:/' + testPath;
+}
+
+function convertFilesStatusIntoBenderFilter( filesStatus ) {
```
I like the sequential flow of this function that it checks for specific types of changes one after another. However, maybe each check can be a separate function, because ATM it's quite longish.

f1ames commented on test/fileToBenderPathTests.js:
```diff
@@ -0,0 +1,105 @@
+
+});
+
+function MakePathTest( changedFiles, expectedPath, exists ) {
```
It's basically the entire test function, so maybe something like `assertPath`, `assertPaths`?

f1ames commented on test/fileToBenderPathTests.js:
```diff
@@ -0,0 +1,105 @@
+	};
+}
+
+function MakeMultipleResultsTest( changedFiles, expectedPaths ) {
```
Maybe we could merge this function with `MakePathTest` to have only one test function? Especially that both are very similar and differ slightly only on assertion level?

f1ames commented on src/diff/diffAnalyzer.js:
```diff
@@ -0,0 +1,81 @@
+		return fileStatus.split('\t');
+	});
+}
+const path = require( 'path' );
```
I would move `require` to the top, then have "public" function (ones which are exported) and then private ones (so `testPathToBenderFilter` for now).

f1ames reviewed (commented):
**This is the first part of the review. I'll get back to it soon :timer_clock:**.

Good start :+1: 

I like the fact that you have added tests :+1: 

Also the code is quite simple and easy to understand which is a big plus :+1: 

---

I somehow miss some short explanation how it works (section in the README could be useful) with sample git output and the way it's processed and what is the final result.

f1ames reviewed (changes requested):
**Part two** :smile: 

Ok, so I have added one (the last I hope) missing piece here - plugins dependency map (see e648024). It was mentioned in #1:

> If plugin code is changed we should run all tests for this plugin and all plugins which depends on it (in this case `embed`, `image2`,...)

The code just scans all `plugins/plugin.js` files and builds dependency map. So if something changes in the plugin on which other plugins rely, the test for all those plugins should be run (see https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12/commits/bbdcf6a4ed1f2bcd65c9ccd78748953279d52c43#diff-c506b66136cfb6afc94b0a714576eddbb041676636a0aebf2d43ee18782a8c1eR78-R82).

Please take a look on my changes and feel free to refactor whether needed. Two unit tests are failing now, but since it will be refactored slightly (see https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622106515) I'm leaving it to you :wink: 

---

Please also remember about proper code style - so using `'` consistently and correct spacing, see https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12/commits/474bd387dfd1ba1561174079d744e751f04082e9.

---

And the last part should be integrating diffing mechanism into runner so somewhere here - https://github.com/ckeditor/ckeditor4-benderjs-runner/blob/ce13c88e4a0bae73aeeb3198fd5b3fcd7c66ea52/src/runner.js#L56. Now it uses value passed from cmd line but it should use diff result instead. And I would assume branch names which will be used for diff will be passed to runner as cmd arguments.

sculpt0r commented:
@f1ames - ready for review.

- rework assertion in tests. Actually, we don't need exactly the same order of filters. This should be helpful to make those tests more elastic to eventually changes in code. Also renamed and refactor assertions according to (https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622104894, https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622106515)
- Split convert function onto smaller parts, due to (https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622097044)
- Fixed failing tests. One of them requires modifying expected results. Because besides direct filter to the modified plugin, we need filters for plugin dependencies. The second one was also failing due to plugin dependencies. The result was not the array of strings, but actually the array of arrays with duplicated filters. It was introduced by mapping plugin deps to filters.
- I added, but then remove nvmrc with Node 14 (due to `flatMap`), but the bender fails to run. So I used `reduce` to flatten results in differ.
- Modified differ, so it is importable as a module from other scripts. Now it returns Promise which resolves with all generated filters or rejects with error from the git process. This also solves problems with output (https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622095126). Also, I gather all chunks of data and resolve only on `close` (https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#discussion_r622092051)
- Added `respository.targetBranch` and `repository.currentBranch` config.
- Integrate differ with the original runner. Configuration is taken from the config file (for branches and ckeditor directory location). This is slightly different from what you said here (https://github.com/ckeditor/ckeditor4-benderjs-runner/pull/12#pullrequestreview-648117043) in the last sentence, about passing branches from the command line.
- Fix TODO you left for changes in core. Now it also generates filters for plugins: `widget`, `dialog` and `clipboard`
- I change how the dependency map is passing to the analyzer. Currently, the map is created in differ based on ckeditor4 directory localization (or just passed as mock in tests) to the analyzer. So the map is creating by a specific environment (test or differ) rather than initialized deep inside the analyzer.
- Added some explanation about usage to README. I wasn't sure if I should also remove the first point from "what's next". But, since there are some mentions about using additional API to detect which branch is targeting in PR, I decide to not touch it. If we decide to merge this one, maybe we should extract the issue for detecting the target branch in PR?
- Fix for code style

sculpt0r commented:
Rebase on newest master. :see_no_evil: I forget to do this right at the beginning...
- merged changes in README (remove filter from example command)
- correct runner, so the browser name is now taken from the second argument (not third) - since I remove filter from command line arguments

f1ames commented on test/fileToBenderPathTests.js:
```diff
@@ -0,0 +1,153 @@
+	return includesAny( actual, expected );
+}
+
+function includesAny( actual, expected ) {
```
Maybe we could use native [Array.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) here?

f1ames replied:
It checks if the actual array includes all the elements from expected one? :thinking:

f1ames replied:
:point\_up: So it should be named `includesAll` maybe?

f1ames replied:
Let's skip this test refactoring for now so we can start using the runner.

f1ames commented on test/fileToBenderPathTests.js:
```diff
@@ -0,0 +1,153 @@
+	// Two arrays have the same length.
+	// They contains the same elements, but not necessary in the same order.
+	// There are no duplicates in actual array
+	if ( actual.length !== expected.length ) {
+		return false;
+	}
+
+	const hasDuplicates = new Set( actual ).size !== actual.length;
+	if ( hasDuplicates ) {
+		return false;
+	}
```
Shouldn't we check for the same length after removing duplicates?

f1ames replied:
Or having duplicates means we don't treat those arrays as the same?

sculpt0r replied:
One of the assumptions is that we have only one filter of any type. So if we found any duplicate in generated filters - we can't accept results.

f1ames replied:
Let's skip this test refactoring for now so we can start using the runner.

f1ames reviewed (commented):
This is the first part of the review. I spent some time integrating this `diff` script into entire workflow so there are some adjustments here and there. Everything is in the https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/pull/3 PR.

The only issue is that on the CI the diff returns empty result, while the same directory structure locally resulted in correct paths for me. Haven't figured out yet what's going on. See https://travis-ci.com/github/ckeditor/ckeditor4-benderjs-runner-tests/builds/225466699#L321-L327.

f1ames commented:
> The only issue is that on the CI the diff returns empty result, while the same directory structure locally resulted in correct paths for me. Haven't figured out yet what's going on. See https://travis-ci.com/github/ckeditor/ckeditor4-benderjs-runner-tests/builds/225466699#L321-L327.

Ok, so there were 2 issues here - branches weren't checked locally (so using `t/1` which points to local branches diffs nothing) and by default Travis does shallow copy (so you are not able to checkout branches anyway).

The first one was used with adding `origin/` to branch names: [https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/blob/acae21dcb69780fb359c99f7389c06282a15d420/.travis.yml#L65-L70](https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/blob/acae21dcb69780fb359c99f7389c06282a15d420/.travis.yml#L65-L70)

The second one by disallowing shallow clone: https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/blob/acae21dcb69780fb359c99f7389c06282a15d420/.travis.yml#L16-L17

Apart from that it works - https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/runs/2685693327 so I'm for merging and starting using it. Then we can iteratively improve it if needed. I will prepare PR for `ckeditor4` repo based on https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/pull/3 too.

f1ames reviewed (approved):
And it's alive :tada: Nice job @sculpt0r :+1: 

All tests were done in full integration with `ckeditor4` repo (it's fork in fact) in this PR - https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/pull/3.

Recently tested full test runs:

* With failing tests: https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/runs/2692224757
* Without failing tests: https://github.com/ckeditor/ckeditor4-benderjs-runner-tests/runs/2692334714 (apart from positioning tests with 56.70px vs 56.69px check :see_no_evil: )
